core/src/test/scala/unit/kafka/server/DeleteTopicsRequestTest.scala

```diff
@@ -0,0 +1,120 @@
+    validateValidDeleteTopicRequests(new DeleteTopicsRequest(Set("topic-3", "topic-4").asJava, timeout))
+  }
+
+  private def validateValidDeleteTopicRequests(request: DeleteTopicsRequest): Unit = {
+    val response = sendDeleteTopicsRequest(request, 0)
+
+    val error = response.errors.values.asScala.find(_ != Errors.NONE)
+    assertTrue(s"There should be no errors, found ${response.errors.asScala}", error.isEmpty)
+
+    request.topics.asScala.foreach { topic =>
+      validateTopicIsDeleted(topic)
+    }
+  }
+
+  @Test
+  def testErrorDeleteTopicRequests() {
+    val timeout = 30000
+    val timeoutTopic = "invalid-timeout"
+
+    // Basic
+    validateErrorDeleteTopicRequests(new DeleteTopicsRequest(Set("invalid-topic").asJava, timeout),
+      Map("invalid-topic" -> Errors.INVALID_TOPIC_EXCEPTION))
+
+    // Partial
+    TestUtils.createTopic(zkUtils, "partial-topic-1", 1, 1, servers)
+    validateErrorDeleteTopicRequests(new DeleteTopicsRequest(Set(
+      "partial-topic-1",
+      "partial-invalid-topic").asJava, timeout),
+      Map(
+        "partial-topic-1" -> Errors.NONE,
+        "partial-invalid-topic" -> Errors.INVALID_TOPIC_EXCEPTION
+      )
+    )
+
+    // Timeout
+    TestUtils.createTopic(zkUtils, timeoutTopic, 5, 2, servers)
+    validateErrorDeleteTopicRequests(new DeleteTopicsRequest(Set(timeoutTopic).asJava, 1),
+      Map(timeoutTopic -> Errors.REQUEST_TIMED_OUT))
+    // The topic should still get deleted eventually
+    TestUtils.waitUntilTrue(() => !servers.head.metadataCache.contains(timeoutTopic), s"Topic $timeoutTopic is never deleted")
+    validateTopicIsDeleted(timeoutTopic)
+  }
+
+  private def validateErrorDeleteTopicRequests(request: DeleteTopicsRequest, expectedResponse: Map[String, Errors]): Unit = {
+    val response = sendDeleteTopicsRequest(request, 0)
+    val errors = response.errors.asScala
+    assertEquals("The response size should match", expectedResponse.size, response.errors.size)
+
+    expectedResponse.foreach { case (topic, expectedError) =>
+      assertEquals("The response error should match", expectedResponse(topic), errors(topic))
+      // If no error validate the topic was deleted
+      if (expectedError == Errors.NONE) {
+        validateTopicIsDeleted(topic)
+      }
+    }
+  }
+
+  @Test
+  def testNotController() {
+    val request = new DeleteTopicsRequest(Set("not-controller").asJava, 1000)
+    val response = sendDeleteTopicsRequest(request, 0, notControllerSocketServer)
+
+    val error = response.errors.asScala.head._2
+    assertEquals("Expected controller error when routed incorrectly",  Errors.NOT_CONTROLLER, error)
+  }
+
+  private def validateTopicIsDeleted(topic: String): Unit = {
+    val metadata = sendMetadataRequest(new MetadataRequest(List(topic).asJava)).topicMetadata.asScala
+    assertFalse(s"The topic $topic should not exist", metadata.exists(p => p.topic.equals(topic) && p.error() == Errors.NONE))
+  }
```

Comment:
Since the propagation of metadata is async to different brokers, we may need to do the test in a waitUntilTrue() block.